guidedremediation/internal/manifest/python/poetry.go

```diff
@@ -0,0 +1,230 @@
+	var reqs []resolve.RequirementVersion
+	if deps == nil {
+		return reqs
+	}
+
+	for _, reqStr := range deps {
+		d, err := pypi.ParseDependency(reqStr)
+		if err != nil {
+			log.Warnf("failed to parse python dependency in pyproject.toml %q: %v", reqStr, err)
+			continue
+		}
+
+		var dt dep.Type
+		if optional {
+			dt.AddAttr(dep.Opt, "")
+		}
+		reqs = append(reqs, resolve.RequirementVersion{
+			VersionKey: resolve.VersionKey{
+				PackageKey: resolve.PackageKey{
+					System: resolve.PyPI,
+					Name:   d.Name,
+				},
+				Version:     d.Constraint,
+				VersionType: resolve.Requirement,
+			},
+			Type: dt,
+		})
+	}
+	return reqs
+}
+
+// Read parses the manifest from the given file.
+func (r poetryReadWriter) Read(path string, fsys scalibrfs.FS) (manifest.Manifest, error) {
+	path = filepath.ToSlash(path)
+	f, err := fsys.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var proj pyProject
+	if _, err := toml.NewDecoder(f).Decode(&proj); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal pyproject.toml: %w", err)
+	}
+
+	allReqs := []resolve.RequirementVersion{}
+	groups := make(map[manifest.RequirementKey][]string)
+
+	// Dependencies
+	allReqs = append(allReqs, parseDependencies(proj.Project.Dependencies, false)...)
+
+	// Optional dependencies
+	for groupName, deps := range proj.Project.OptionalDependencies {
+		groupReqs := parseDependencies(deps, true)
+		allReqs = append(allReqs, groupReqs...)
+		for _, r := range groupReqs {
+			key := manifest.RequirementKey(r.VersionKey.PackageKey)
+			groups[key] = append(groups[key], groupName)
+		}
+	}
+
+	return &pythonManifest{
+		filePath: path,
+		root: resolve.Version{
+			VersionKey: resolve.VersionKey{
+				PackageKey: resolve.PackageKey{
+					System: resolve.PyPI,
+					Name:   proj.Project.Name,
+				},
+				VersionType: resolve.Concrete,
+				Version:     proj.Project.Version,
+			},
+		},
+		requirements: allReqs,
+		groups:       groups,
+	}, nil
+}
+
+// Write writes the manifest after applying the patches to outputPath.
+func (r poetryReadWriter) Write(original manifest.Manifest, fsys scalibrfs.FS, patches []result.Patch, outputPath string) error {
+	return write(fsys, original.FilePath(), outputPath, patches, updatePyproject)
+}
+
+// updatePyproject takes an io.Reader representing the pyproject.toml file
+// and a map of package names to their new version constraints, returns the
+// file with the updated requirements as a string.
+func updatePyproject(reader io.Reader, requirements map[string][]VersionConstraint) (string, error) {
+	data, err := io.ReadAll(reader)
+	if err != nil {
+		return "", fmt.Errorf("error reading requirements: %w", err)
+	}
+
+	var sb strings.Builder
+	inProjectSection := false
+	inOptionalDepsSection := false
+	inDepArray := false
+	for _, line := range strings.SplitAfter(string(data), "\n") {
+		trimmedLine := strings.TrimSpace(line)
```

Comment:
We probably also want to handle toml comments.
But there's also the edge case where a `#` is within a string.